### DISCUSSION TITLE
When can I use a Reducer? What does it emit?

### DISCUSSION THREAD

**Discussion by spagdoon0411** [awslabs/project-lakechain#51] [Q&A] [answered]
Please let me know if there's a resource I missed that might answer this! 

To start, [here's the Reducer documentation](https://awslabs.github.io/project-lakechain/flow-control/reducer/). 

From the documentation, I wasn't able to get a complete understanding of when a Reducer can or can't be used prior to another node. 

For instance, say I'm trying to split a slideshow into images, then give them all to Claude and ask for a summary of the presentation. The Reducer documentation states this node allows us to "group relevant documents into a single semantical envelope and perform combined operations on them." Broadly, that _seems_ right for this use case. 

But I'm also doubtful of whether that's possible: it doesn't look like there's a spot to pass a funclet to an `AnthropicTextProcessor` `Builder` that defines how to perform the reduction correctly for Claude's input. (In contrast, the podcast generator example features a `withIntent` call on an `FfmpegProcessor` `Builder` that takes a concat funclet; see [line 231 here](https://github.com/awslabs/project-lakechain/blob/main/examples/end-to-end-use-cases/building-a-podcast-generator/stack.ts)). 

```typescript
// ...

import { concat } from './funclets/ffmpeg';

// ...

const ffmpeg = new FfmpegProcessor.Builder()
      .withScope(this)
      .withIdentifier('FfmpegProcessor')
      .withCacheStorage(cache)
      .withVpc(vpc)
      .withSource(voiceReducer)
      .withIntent(concat)
      .build();
```

Information that might help is what CloudEvent(s) is/are emitted into SNS. (E.g., the following diagram notes "aggregated events"--but in what sense are they aggregated and how are the CloudEvents logically organized?)

![image](https://github.com/user-attachments/assets/85efe9a3-4037-4afd-96dc-db78e78cbbdb)

**Comment by HQarroum** [comment, marked as answer]
Hey @spagdoon0411!

That's a good question. The **TLDR;** is that the reducer aggregates the CloudEvents passed by upstream middlewares and combines them into a JSON document (a JSON array, in fact) that contains the aggregated cloudevents. It does that aggregation based on the given reducer strategy (TimeWindowStrategy, StaticCounterStrategy, ConditionalStrategy). The output of a reducer is always `application/cloudevents+json`. So the output of a reducer would like like this :

```
[{
  "specversion": "1.0",
  "id": "1780d5de-fd6f-4530-98d7-82ebee85ea39",
  "type": "document-created",
  ...
}, {
  // Another cloudevent
}]
```

Now regarding your question of when a reducer can be used prior to another middleware is it depends on whether that middleware supports `application/cloudevents+json` as an input type. Today, only the following middlewares that support composite inputs are the following :
- [Anthropic Text Processor](https://awslabs.github.io/project-lakechain/generative-ai/anthropic-text-processor/)
- [Structured Entity Extractor](https://awslabs.github.io/project-lakechain/generative-ai/structured-entity-extractor/)
- [FFMPEG Processor](https://awslabs.github.io/project-lakechain/video-processing/ffmpeg-processor/)
- [Zip Processor](https://awslabs.github.io/project-lakechain/archive-processing/zip-deflate-processor/)
- [Tar Inflate Processor](https://awslabs.github.io/project-lakechain/archive-processing/tar-inflate-processor/)

Hope I answered your question.

**Reply by spagdoon0411** [reply]
Hi Halim! The doc page _does_ list `application/cloudevents+json` as an output MIME type; I probably could have paid closer attention to that. 

I also saw you were working on a guide for custom middleware modules--I think defining complex reducer behavior would be a great use case. 

Thanks for the answer!
